client/py_client/ipfs.py

```diff
@@ -15,18 +17,43 @@
 class Ipfs:
     """ Minimal wrapper for the ipfs cli """
     @staticmethod
-    def add(path_to_files, local = False):
+    def add_recursive(path_to_files, local = False):
         if not (use_ipfs_gateway or local):
             return "QmP2fzfikh7VqTu8pvzd2G2vAd4eK7EaazXTEgqGN6AWoD"
-        ret = ''
         if local:
-            ret = subprocess.run(["ipfs", "add", path_to_files], stdout=subprocess.PIPE)
+            ret = subprocess.run(["ipfs", "add", "-rw", path_to_files], stdout=subprocess.PIPE)
+            return take_only_last_cid(ret)
         else:
-            ret = subprocess.run(["curl", "-X", "POST", "-F", f"file=@{path_to_files}", "-u", ipfs_api_key, ipfs_add_url], stdout=subprocess.PIPE)
-        return take_only_last_cid(ret)
+            headers = { }
+            params = ()
+            if os.path.isdir(path_to_files):
+                params = (
+                    ('pin', 'true'),
+                    ('recursive', 'true'),
+                    ('wrap-with-directory', 'true'),
+                )
+            else:
+                params = (
+                    ('pin', 'true'),
+                )
+            files = generate_file_list(path_to_files)
+            auth = ipfs_api_key.split(":")
+            response = requests.post('https://ipfs.infura.io:5001/api/v0/add', headers=headers, params=params, files=files, auth=(auth[0], auth[1]))
```

Comment:
Shouldn't we make this configurable to also use the encointer end-point?

Reply:
No, this is how we add a file to ipfs using the infura api with the user -u argument.
When a file is added to an ipfs node, it will notify the other nodes that it has the content with cid xxx.
So uploading like this will still allow us to view the file with our encointer gateway.
I'm dont remember exactly why we actually use infura, i guess its because by specifying a user & password it uploads the files to our node,  and nobody can upload to our node without knowing our user&password. So we are protected from attacks that fill up our space on the node

Reply:
there is too much confusion here about our ipfs strategy. Please leave as-is for now. We will fix this in January

Reply:
ok, but for the record, this would be a working PR that works also on the app side. without this, we don't fetch any new icons

Reply:
we can merge this if @clangenb agrees. ipfs fixe can and should be its own PR

Reply:
> I have the feeling some parts of the code is more complex than necessary, but I have no domain knowledge to give better suggestions. :)
> 
> I guess this one is obsolete then? [encointer/encointer-wallet-flutter#249](https://github.com/encointer/encointer-wallet-flutter/pull/249)

yes, this would make the other PR on the app side obsolete. 
And I agree, the implementation looks more complex then it should, and I hoped it would be easier applieable, but it didn't work the easy way. I believe the ipfs http client for python should be an abstraction level higher and a better wrapper for this implementation: https://github.com/ipfs-shipyard/py-ipfs-http-client, but when I tried it a cuple of months ago i didn't succeed and didn't try it ever since. 
There is for infura a python tutorial, and they use the response library (which I also use in this PR):  
https://blog.infura.io/getting-started-with-infuras-ipfs-service-updated-june-2021/
(If you're interested to see how it would looks like with the curl command, you can see the commented code in my first commit)